### ISSUE TITLE
Demo reporting wrong teams

### ISSUE THREAD

**Issue by LarsKoelpin** [markus-wa/cs-demo-minifier#41]
I noticed, that analyszed faceit demos (faceit GoTV Demo) report wrong teams

DemoLink: https://demos-europe-west2.faceit-cdn.net/csgo/3f6aa669-1f03-4009-946b-60defd8c7703.dem.gz

Here is what I got:
(Notice) Team 0 on LRSK, tkausla)

```json
0: {"id": 8, "name": "TukezuX", team: 2, npc: null}
1: {"id": 1, "name": "GOTV", team: 0, npc: true}
2: {"id": 6, "name": "AleksiDp", team: 2, npc: null}
3: {"id": 4, "name": "Derputz", team: 3, npc: null}
4: {"id": 5, "name": "suvkkijesh", team: 2, npc: null}
5: {"id": 7, "name": "YoungGo0", team: 2, npc: null}
6: {"id": 9, "name": "BOTYTH", team: 3, npc: null}
7: {"id": 2, "name": "Jesukii", team: 2, npc: null}
8: {"id": 3, "name": "jxns", team: 3, npc: null}
9: {"id": 10, "name": "LRSK", team: 0, npc: null}
10: {"id": 11, "name": "tkausla", team: 0, npc: null}
```

Here is what I expect
```json
0: {"id": 8, "name": "TukezuX", team: 2, npc: null}
1: {"id": 1,  "name": "GOTV", team: 0, npc: true}
2: {"id": 6,  "name": "AleksiDp", team: 2, npc: null}
3: {"id": 4,  "name": "Derputz", team: 3, npc: null}
4: {"id": 5,  "name": "suvkkijesh", team: 2, npc: null}
5: {"id": 7,  "name": "YoungGo0", team: 2, npc: null}
6: {"id": 9,  "name": "BOTYTH", team: 3, npc: null}
7: {"id": 2,  "name": "Jesukii", team: 2, npc: null}
8: {"id": 3,  "name": "jxns", team: 3, npc: null}
9: {"id": 10,  "name": "LRSK", team: 3, npc: null}
10: {"id": 11,  "name": "tkausla", team: 3, npc: null}
```

Do you know where the issue could come from?

Maybe I'm mistaken and there is some kind of team change event?

**Comment by LarsKoelpin** [comment]
I know, on Another issue, we talked about effiency and usability.

But, as the swap team may come in every frame, should we maybe serialize the team in each snapshot and remove the "teams" header completely? Also the swap_teams event could be ommited then, I guess?

I'm thinking in terms of usage here, as usually I do render every snapshot in every frame in my own tool. I want to avoid to detect changes and instead think more of 
a function which renders 

snapshot -> UI

**Comment by markus-wa** [comment]
events are accurate to the tick, so you should be able to iterate through ticks and apply events as soon as the snapshot tick is greater than the `ticks[n].nr`. that way you don't need that data on each snapshot.

**Comment by LarsKoelpin** [comment]
That's exactly what I was arguing against.

The Client of the JSON has now to figure out who is on which team on tick 15001 by analysing all events which occured one after one. It would be much easier to know "THIS IS THE GAME STATE; deal with it"

But I'm fine either way. Don't know if this needs a fix related to the original issue. As swap team events are fired correctly. (Although the JSON-header-teams-section imo is some kind of misleading).

I guess we could close the issue then

**Comment by markus-wa** [comment]
Well you already need this logic for all game events anyway. Sure we can keep duplicating things into snapshots but the more we do this thr slower it will become and the more bloated it will become.

Sure the team field is a small amount of data but I think it's more about the strategy going forward - and adding everything there just isn't going to scale.

You don't have to analyse all preceding events if we add oldTeam and newTeam onto `team_swap`. Maybe i should have been more clear on this, I know I'm often sort of rambling and posting really short responses (partially due to a lack of time) so sorry about that.😅

I'm tempted to agree on removing team header and just having team switch events though with appropriate info. Seems like the headers would be useless then. But thlat would require wider ranging changes. Might have another think about it.

In short: we'll need to add newTeam and oldTeam onto that switch event. Would that solve it for you?

**Comment by markus-wa** [comment]
I'll try to give this all some more thought soon, there's a bunch of stuff I'm struggling to put into words without some more time but my gut feeling says this is the right way to go.

**Comment by LarsKoelpin** [comment]
Hey,
```
 Well you already need this logic for all game events anyway. 
```
That's were I'm comming from. It's a pain to work with (just talking for my usecase, a demoplayer, I could imagine is one of  the "main usecase"s).

 My usecase is, I want to somehow render the "demo" using json format lets say in a textual style. If I'm using the resulting JSON in that way: Here is my tick/frame whatever: Draw each of these things: this is my score, this is my hp, this is my position(where we actually calculate the current state i guess and do not output the delta). 

```
Sure we can keep duplicating things into snapshots but the more we do this thr slower it will become and the more bloated it will become.
```
yep. That's why I am saying its more about usablity vs efficiency. If efficiency is your point, I'll take it :)

```
You don't have to analyse all preceding events if we add oldTeam and newTeam onto team_swap. Maybe i should have been more clear on this, I know I'm often sort of rambling and posting really short responses (partially due to a lack of time) so sorry about that.
```
Image the usecase from above: I continously have to check:
* whats the snapshot I'm presenting (lets say index 6 in the array)
* at which tick index are we (lets say index 4), 
* which events are happening, were they applied to the UI already? (like a grenade explosion, or teamchange)
* which events to I have to apply to the current state (Is the explision rendered already?)
* lets say events at index 4 contains: team_change
* I have to check:
  * Do events and snapshots-ticks align -> Yes
  * events contain team_swap
  * **update some kind of temporary state, which merges snapshots and events to get the truth about the gamestate and keeps track of the state at runtime** (lets call this variable (EntityIdToCurrentTeam)**
* If I want to know the current team for a player: EntityIdToCurrentTeam[currentPlayer.id]

So: I have to re-derive the real game state by "merging current state with the events " happening and sort of reimplementing the logic

IIRC this is how games and videos actually work:

Each Frame:
* This is your team  + color
* This is your posiition
* These grenades are active 
* This is your equipment
etc..

As of writing this, I realize that maybe I'm too much "thinking in states" (that's how react.js works, if you are not into javascript), but these are my Painpoints in a nutshell. 
Maybe it should not be part of this library and I'm too much into my own problems. 

Either way, there are definately usecases where someone e.g. wants to analyse "game events" instead, where the current structure may be way superior to the "snapshot"-contains-everything-approach I presented.

I don't know. I guess it's fine and we should keep it as it is :)

**Comment by markus-wa** [comment]
I can see and understand your pain. As an advocate of functional programming all these state transitions are far from beautiful.

```
IIRC this is how games and videos actually work:

Each Frame:

    This is your team + color
    This is your posiition
    These grenades are active
    This is your equipment
    etc..
```

Internally yes, but over the network and in demos
Csgo actually does events and property updates, which is where I got the idea from. Sending all data to the client every tick just wouldn't be feasible.

But I will give this all a big think. I think you are certainly onto something here.

**Comment by LarsKoelpin** [comment]
related to the original issue: 

I just noticed there was a team attribute present in the entityUpdates. It was just never filled (always nil, therefore not rendered to json). (commit 02e4c810837b617e3e4fd8ad0fde327dc32c6703)

**Comment by markus-wa** [comment]
the crazy things i get up to! nice catch.

still not sure what we wanna do with it long term but i guess might as well activate it if it's there.

I'd say there's a good chance it wouldn't survive until v1.0.0 (if that ever happens) but that's an issue for another day.

**Comment by markus-wa** [comment]
could you try this? https://github.com/markus-wa/cs-demo-minifier/pull/43

**Comment by LarsKoelpin** [comment]
Does the trick and works on my machine
